ArgumentCaptor struct and tests

tylersammann opened the pull request:
Resolves Issue: #262 

- added `ArgumentCaptor` struct that composes `Matcher`
- added constructors for `ArgumentCaptor`
- included unit tests for `ArgumentCaptor` methods

The `Captor()` method takes a `Matcher` interface to make it as flexible as possible. `AnyCaptor()` is just there because it seems like the `anyMatcher` is probably going to be used with captors the most.

poy commented:
What is the advantage of this over just using `.Do()`? This adds quite a bit to the existing API.

poy commented:
Also, I would like to say, I like the idea of gomock being able to generate "spies" in some way, but I want to ensure the API is solid.

tylersammann commented:
@poy gotcha. I wrote a little bit about using `.Do()` here. https://github.com/golang/mock/issues/262#issuecomment-462901093 
Mostly it just requires a lot of custom code to set up anything that stores arguments for later inspection, especially if you want to inspect multiple arguments. Also, in the multiple arg senario I find that it is nice to use the named captors rather than relying on argument order or re-implementing everything in the `.Do()` clause for each situation. 

In terms of API change size, I think I'm only adding one 37 line file to the API itself. https://github.com/golang/mock/blob/bdfd0e24d230faaafd3c9d815aab7c1266ce9685/gomock/argument_captor.go The rest of the additions are testing or sample code related. (plus one `.gitignore` change that we can scrap if you don't want that)

I am happy to discuss changes to the API though if you can see places to improve it! My team has been using my forked ArgumentCaptors for the last six months with a lot of success and I'd love to see them added to the main branch.

nguyenfilip commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,37 @@
+	return ac.Matcher.Matches(x)
+}
+
+// If called multiple times, return the last argument value used
```
Maybe just rename to LastValue() and drop the comment? Also the comment is a bit misleading. It says "if called multiple times" but it doesn't refer to calling of this Value() method.

codyoss replied:
I think either way is a tradeoff here, `LastValue` vs `Value`. I would actually lean towards the original implementation of `Value` /w the documentation.  @cvgw WDYT?

tylersammann replied:
I would also be ok to go back to `Value()` and `AllValues()`.

nguyenfilip replied:
Please feel free to go back to the original implementation as @codyoss suggested.

tylersammann replied:
alright I'll do that

nguyenfilip commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,37 @@
+package gomock
+
+type ArgumentCaptor interface {
```
There is no need to export the ArgumentCaptor interface. Reasons:
1. Exported interface means we never can add or remove methods from it without breaking backwards compatibilty
2. Even in the this PR, there is no practical reason for the interface. It has only one implementation

Instead I would argue just to export the `ArgumentCaptor` struct and make `Captor()` method return `*ArgumentCaptor`

nguyenfilip commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,37 @@
+	return &argumentCaptor{Matcher: m}
+}
+
+func AnyCaptor() ArgumentCaptor {
```
Maybe rename to Any()? ArgumentCaptor.AnyCaptor() stutters

tylersammann replied:
the `gomock` package already has a public `Any()` method, so I can't use that name.

nguyenfilip commented:
@codyoss the naming of ArgumentCaptor seems fine to me. Any other suggestions?

tylersammann commented on gomock/argument_captor_test.go:
```diff
@@ -0,0 +1,128 @@
+
+	anyCaptor := gomock.AnyCaptor()
+
+	if !anyCaptor.Matches(192394) {
```
There's a pattern that every source `.go` file in the lib has its own set of unit tests, so I was conforming to that. Also, I like the idea of the tests being specific to this new feature.

codyoss commented:
Thoughts on just calling it `Captor` vs `ArgumentCaptor`? The latter seems more Java like to me,  but it is nice an explicit.

codyoss commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,40 @@
+
+// ArgumentCaptor is a struct that composes a Matcher, and extends it by storing given arguments in the values slice.
+type ArgumentCaptor struct {
+	Matcher
```
I think `m Matcher` would suit us better here. If you ever need to do anything fancy /w reflection or anything working with embedded fields can be hard.

codyoss commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+// AnyCaptor is a helper method that returns a new *ArgumentCaptor struct with the matcher set to an anyMatcher
+func AnyCaptor() *ArgumentCaptor {
```
Thoughts on `CaptureAny()`?

tylersammann replied:
This would be ok with me if the call usually happened in the `EXPECT` line, but because captors aren't that useful unless they're declared in advance, I think it makes more sense to keep it as is and view it like a constructor. 

Here's an example:
```go
// here we don't have a way to grab the values
mockDao.EXPECT().InsertIDs(gomock.CaptureAny()).Times(1)

// we need to store the captor first 
idCaptor := gomock.AnyCaptor()
mockDao.EXPECT().InsertIDs(idCaptor).Times(1)
actualIDs := idCaptor.LastValue().([]int)
```

codyoss commented on sample/captor/service_test.go:
```diff
@@ -0,0 +1,31 @@
+package captor
```
Can you please move this to `mockgen/internal/tests`.  This sample folder should probably never have been located where it is. As is this technically leaks into the public api.

codyoss commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,40 @@
+package gomock
```
Please add license to all of the non-generated files.

codyoss commented:
@tylersammann Thanks for working on this!

tylersammann commented:
@codyoss I like the name `ArgumentCaptor` because it is explicit. `Captor` isn't too bad though. I'm up to change it if others agree.

tylersammann commented:
@nguyenfilip @codyoss I think I've addressed most of the PR comments and change requests. Let me know how you feel re: `Value()` vs. `LastValue()` and I'll update accordingly.

codyoss commented:
Thanks, sorry for the delay. I will get back to review this by the end of the week.

nguyenfilip commented:
@tylersammann I don't have any more comments. I will let @codyoss approve because there is stil the discussion open about naming.

tylersammann commented:
@nguyenfilip @codyoss this should be ready for another look. I've reverted the method names back to `Value` and `AllValues`

codyoss commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,59 @@
+	return ac.values[len(ac.values)-1]
+}
+
+// AllValues returns the all arguments the matcher was called with as a []interface{}.
```
Reword similarly to how Value is worded.

codyoss commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// Captor is a helper method that returns a new *ArgumentCaptor struct with Matcher set to the given matcher m
+func Captor(m Matcher) *ArgumentCaptor {
```
Reword: "Captor returns an ArgumentCaptor..." and explain how the matcher is used. Should not reference the private field m.

codyoss commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// AnyCaptor is a helper method that returns a new *ArgumentCaptor struct with the matcher set to an anyMatcher
+func AnyCaptor() *ArgumentCaptor {
```
Reword: "AnyCaptor returns an ArgumentCaptor that will..." explain that it will accept any input.

codyoss commented on gomock/argument_captor_test.go:
```diff
@@ -0,0 +1,142 @@
+	captor.Matches(intArg)
+	captor.Matches(sliceArg)
+
+	actualValue := captor.Value().([]string)
```
Given the test, I know this will work, but lets check the type and error if it does not match. Will be useful if this test ever got refactored.

codyoss commented on mockgen/internal/tests/captor/dao.go:
```diff
@@ -0,0 +1,19 @@
+package captor
```
License

codyoss commented on gomock/argument_captor.go:
```diff
@@ -0,0 +1,59 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gomock
```
Would you add a couple examples of both a simple and more complex case using this as an end-user.

codyoss commented:
@tylersammann Are you still interested in this change?

tylersammann commented:
Yep definitely @codyoss. I can address your review this weekend

codyoss commented:
No worries, I still owe you another through review too. I apologize for not being more responsive here. These have been unprecedented times.

tylersammann commented:
I'll wait for you to review again before I push any changes. Thanks for looking @codyoss

tylersammann commented:
@codyoss I came back through and addressed your first review because I had some time. Feel free to take another look!

hgl commented:
Thanks everyone for the hard work.

Any progress on this PR?

isopropylcyanide commented:
Thanks folks. Can we get this merged? This'd be a great feature add.

tylersammann commented:
I'm happy to continue on this. It'll need a re-review from @codyoss and it's been a while, so a few tests and a merge conflict needs fixing.

codyoss commented:
@tylersammann  Apologies I dropped the ball on this PR. Let me take a look again here this week. One thing to consider for the future now is also generics with 1.18 coming soon. I would like to think about if generics might change surface. I will get back to you soon!

tylersammann commented:
@codyoss thanks! I agree, generics might remove the need for type casting captured values, and that would be a big win.

codyoss commented:
I think I am going to close this for now. I believe there might be a better surface design for this given generics now. Let's keep the chat going in the issue though.
